### PR TITLE
Add version command.

### PR DESCRIPTION
**Pull request by sebpuetz**
Could be nice to set something in CI to fail release-builds to keep the version in sync.

### CODE REVIEW

File: finalfrontier-utils/src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,6 @@
+#[macro_use]
+extern crate clap;
```
Comment (danieldk) [inline finalfrontier-utils/src/lib.rs]: This shouldn't be necessary. In `util`, you should just be able to do `use clap::crate_version`.

**Comment by danieldk** [comment]
Oh, and the version should also be added to the metadata of the written embeddings.

**Comment by sebpuetz** [comment]
https://docs.rs/vergen/3.0.4/vergen/ could be something we could use together with `env_option!`

**Comment by sebpuetz** [comment]
Really unsure whether this is any good, but I added a build dependency on `git2` and set `rustc_env` flags through `build.rs` scripts. That way we get info such as the current commit, whether the repo was in dirty state and commits since the last tag.

**Comment by sebpuetz** [comment]
> Really unsure whether this is any good, but I added a build dependency on `git2` and set `rustc_env` flags through `build.rs` scripts. That way we get info such as the current commit, whether the repo was in dirty state and commits since the last tag.

I just spent 3 hours debugging why `git2` was throwing errors about unknown objects before finding out that travis simply didn't clone it because of `git clone --depth=50`. I hope that setting `depth` to false can override this...

**Comment by danieldk** [comment]
> https://docs.rs/vergen/3.0.4/vergen/ could be something we could use together with `env_option!`

That indeed looks interesting! May be an easier way to get the commit sha.

**Comment by sebpuetz** [comment]
iirc, vergen calls `git` directly and parses its output. I think going through `libgit2` bindings is preferable since we don't get external build dependencies. The Rust bindings for `libgit2` are maintained in the `rust-lang` repo and I think are used in cargo https://github.com/rust-lang/git2-rs.

File: finalfrontier-utils/build.rs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            .expect("Could not get description for git repo.");
+
+        let desc = describe
+            .format(Some(
+                DescribeFormatOptions::new()
+                    .always_use_long_format(true)
+                    .dirty_suffix("-dirty"),
+            ))
+            .expect("Could not format description for git repo.");
+        println!("cargo:rustc-env=FINALFRONTIER_GIT_DESC={}", desc);
```
Comment (danieldk) [inline finalfrontier-utils/build.rs]: Is this used in `-utils`?

Comment (sebpuetz) [reply finalfrontier-utils/build.rs]: Don't think so, that's probably copy-pasted from the library's `build.rs`.

**Comment by sebpuetz** [comment]
I amended the commit with a bump in the min Rust version for travis and fixed the print in `-utils`.

I'm still not sold on having the `chrono` dependency and timestamp in `finalfrontier` instead of `finalfrontier-utils`. I think it'd be a better fit to have some struct in `finalfrontier-utils` capturing the start and end timestamps for training, the `finalfrontier-utils` version, the corpus and the number of threads.

Then in `finalfrontier` we could just include the version + git sha. What's your take?

**Comment by danieldk** [comment]
I agree, but I was toying with the idea of merging `finalfusion` and `finalfusion-utils` into one crate. Most uses will be as a program crate and this will have several benefits: the compiler will tell us about dead code that just happens to be `pub` (like the other PR where I removed some subword indices function); we don't need to maintain duplicate dependency lists; makes uploading finalfrontier easier. For the small number of users that want to use finalfrontier as a library externally, that would still be possible AFAIK. I have been using `sticker-utils` as a library in `sticker-python` without any problems.

The `finalfrontier`/`finalfrontier-utils` split stems from when you still needed `finalfrontier` as a library to load the actual embeddings.

I'll make the PR for this merge and then we can see if we want to proceed with this.

**Comment by sebpuetz** [comment]
I think there's the upside that managing dependencies and our API becomes simpler but we're also losing the verification that API changes don't break external training implementations. Although, if the fundamental parts like the `Trainer` and `Vocab` traits and the `SGD` struct are part of the public API, we should get warnings if anything private is leaked.

**Comment by sebpuetz** [comment]
I rebased this on master and removed the `chrono` stuff. I wasn't really happy with bundling that together with the version info. But I'm not sure whether bundling it with `corpus` and `n_threads` makes more sense. Anyways, I think we can consider this in a follow up PR where we could add some interface to extend the meta information in `write_model_binary`.

**Comment by sebpuetz** [comment]
~~~
target/release/ff-train-skipgram tdz-text.conll throwaway --threads 14 --dims 10 --epochs 5 --context 10 -V
ff-train-skipgram 0.6.0 34460ea-dirty
~~~

I removed the difference between `-V` and `--version` since I think this is fine now as a version.

**Pull request merged**